### ISSUE TITLE
error building on mac, member  reference base type 'Span' is not a structure or union

### ISSUE THREAD

**Issue by reneix** [dmlc/xgboost#9667]
I'm building xgboost from source on mac 10.15, getting another error: member reference base type 'Span' is not a structure or union

please help, error details below

system: 10.15
xgboost: 2.0.0
dmlc-core:0.6.0
Apple clang version 12.0.0 (clang-1200.0.32.29)

(base) zhaochao@zhaochaodeMacBook-Pro build % make -j4                                                              
Consolidate compiler generated dependencies of target dmlc
[  1%] Building CXX object dmlc-core/CMakeFiles/dmlc.dir/src/data.cc.o
[  2%] Linking CXX static library libdmlc.a
[ 11%] Built target dmlc
[ 15%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api_error.cc.o
[ 15%] Building CXX object src/CMakeFiles/objxgboost.dir/collective/communicator.cc.o
[ 15%] Building CXX object src/CMakeFiles/objxgboost.dir/collective/in_memory_communicator.cc.o
[ 15%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o
[ 16%] Building CXX object src/CMakeFiles/objxgboost.dir/collective/in_memory_handler.cc.o
[ 17%] Building CXX object src/CMakeFiles/objxgboost.dir/collective/socket.cc.o
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/c_api.cc:29:
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/simple_dmatrix.h:16:
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/gradient_index.h:19:
/Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/../common/ref_resource_view.h:79:32: error: member
      reference base type 'Span' is not a structure or union
    return Span{data(), size()}.size_bytes();
           ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/c_api.cc:6:
In file included from /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:643:
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/memory:3559:7: error: exception specification of
      overriding function is more lax than base version
class __shared_ptr_emplace
      ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/memory:4405:26: note: in instantiation of template
      class 'std::__1::__shared_ptr_emplace<xgboost::common::MmapResource,
      std::__1::allocator<xgboost::common::MmapResource> >' requested here
    ::new(__hold2.get()) _CntrlBlk(__a2, _VSTD::forward<_Args>(__args)...);
                         ^
/Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../common/io.h:387:40: note: in instantiation of function template
      specialization 'std::__1::make_shared<xgboost::common::MmapResource, std::__1::basic_string<char> &, unsigned
      long &, unsigned long &>' requested here
      : AlignedResourceReadStream{std::make_shared<MmapResource>(path, offset, length)} {}
                                       ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/memory:3468:13: note: overridden virtual function is
      here
    virtual ~__shared_weak_count();
            ^
[ 18%] Building CXX object src/CMakeFiles/objxgboost.dir/common/charconv.cc.o
2 errors generated.
make[2]: *** [src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [src/CMakeFiles/objxgboost.dir/all] Error 2
make: *** [all] Error 2

**Comment by hliuca** [comment]
tried on,

mac 14.0
clang 15.0

everything is fine. the command line uses -std=c++17 (make VERBOSE=1). your system (os and compiler) is a bit old?

**Comment by reneix** [comment]
would you please show full make cmd please? thanks

> tried on,
> 
> mac 14.0 clang 15.0
> 
> everything is fine. the command line uses -std=c++17 (make VERBOSE=1). your system (os and compiler) is a bit old?

**Comment by hliuca** [comment]
cd /Users/ericliu/xgboost/build/src && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DDMLC_CORE_USE_CMAKE -DDMLC_LOG_CUSTOMIZE=1 -DDMLC_USE_CXX11=1 -DXGBOOST_BUILTIN_PREFETCH_PRESENT=1 -D_MWAITXINTRIN_H_INCLUDED -D__USE_XOPEN2K8 -I/Users/ericliu/xgboost/include -I/Users/ericliu/xgboost/dmlc-core/include -I/Users/ericliu/xgboost/rabit/include -I/Users/ericliu/xgboost/build/dmlc-core/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fPIC -funroll-loops -Xpreprocessor -fopenmp -I/opt/local/include/libomp -std=gnu++17 -MD -MT src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -MF CMakeFiles/objxgboost.dir/c_api/c_api.cc.o.d -o CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -c /Users/ericliu/xgboost/src/c_api/c_api.cc

**Comment by hliuca** [comment]
v2.0.0 is similar. 544/545 tests passed, and 1 skipped, Socket.Basic.

**Comment by reneix** [comment]
here is my, the -I param seems identical, will find a new env to try...

cd /Users/zhaochao/Downloads/xgboost-2.0.0/build/src && /Library/Developer/CommandLineTools/usr/bin/c++ -DDMLC_CORE_USE_CMAKE -DDMLC_LOG_CUSTOMIZE=1 -DDMLC_USE_CXX11=1 -DXGBOOST_BUILTIN_PREFETCH_PRESENT=1 -DXGBOOST_MM_PREFETCH_PRESENT=1 -D_MWAITXINTRIN_H_INCLUDED -D__USE_XOPEN2K8 -I/Users/zhaochao/Downloads/xgboost-2.0.0/include -I/Users/zhaochao/Downloads/xgboost-2.0.0/dmlc-core/include -I/Users/zhaochao/Downloads/xgboost-2.0.0/rabit/include -I/Users/zhaochao/Downloads/xgboost-2.0.0/build/dmlc-core/include -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.15 -fPIC -funroll-loops -Xpreprocessor -fopenmp -I/usr/local/opt/libomp/include -std=gnu++17 -MD -MT src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -MF CMakeFiles/objxgboost.dir/c_api/c_api.cc.o.d -o CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -c /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/c_api.cc
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/c_api.cc:29:
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/simple_dmatrix.h:16:
In file included from /Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/gradient_index.h:19:
/Users/zhaochao/Downloads/xgboost-2.0.0/src/c_api/../data/../common/ref_resource_view.h:79:32: error: member
      reference base type 'Span' is not a structure or union
    return Span{data(), size()}.size_bytes();
           ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~



> cd /Users/ericliu/xgboost/build/src && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DDMLC_CORE_USE_CMAKE -DDMLC_LOG_CUSTOMIZE=1 -DDMLC_USE_CXX11=1 -DXGBOOST_BUILTIN_PREFETCH_PRESENT=1 -D_MWAITXINTRIN_H_INCLUDED -D__USE_XOPEN2K8 -I/Users/ericliu/xgboost/include -I/Users/ericliu/xgboost/dmlc-core/include -I/Users/ericliu/xgboost/rabit/include -I/Users/ericliu/xgboost/build/dmlc-core/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fPIC -funroll-loops -Xpreprocessor -fopenmp -I/opt/local/include/libomp -std=gnu++17 -MD -MT src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -MF CMakeFiles/objxgboost.dir/c_api/c_api.cc.o.d -o CMakeFiles/objxgboost.dir/c_api/c_api.cc.o -c /Users/ericliu/xgboost/src/c_api/c_api.cc
> 
> cd /Users/ericliu/xgboost/build/dmlc-core && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DDMLC_CORE_USE_CMAKE -DDMLC_USE_AZURE=0 -DDMLC_USE_CXX11=1 -DDMLC_USE_HDFS=0 -DDMLC_USE_S3=0 -DXGBOOST_BUILTIN_PREFETCH_PRESENT=1 -D_DARWIN_C_SOURCE -D_POSIX_C_SOURCE=200809L -D_POSIX_SOURCE -D_XOPEN_SOURCE=700 -D__USE_XOPEN2K8 -I/Users/ericliu/xgboost/dmlc-core/include -I/Users/ericliu/xgboost/build/dmlc-core/include -O3 -DNDEBUG -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -fPIC -msse2 -Wall -Wno-unknown-pragmas -fPIC -O3 -funroll-loops -Xpreprocessor -fopenmp -I/opt/local/include/libomp -std=c++17 -MD -MT dmlc-core/CMakeFiles/dmlc.dir/src/io/line_split.cc.o -MF CMakeFiles/dmlc.dir/src/io/line_split.cc.o.d -o CMakeFiles/dmlc.dir/src/io/line_split.cc.o -c /Users/ericliu/xgboost/dmlc-core/src/io/line_split.cc
> 
> dmlc-core uses -std=c++17, and xgboost uses -std=gnu++17.
> 
> mac 14.0 clang 15.0, and master branch.
> 
> 583 tests from 129 test suites. 581 passed, and 2 tests skipped: SocketTest.Basic and SocketTest.Bind.

**Comment by hcho3** [comment]
Fixed in #9684 and https://github.com/dmlc/xgboost/pull/9693
